File: leet_code/trees/symmetric_tree.js
const inorder = root => {
    if (!root) {
        return [];
    } else {
        const res = [];
        let depth = 0;

        const traverse = (node, depth) => {
            if (node.left) traverse(node.left, depth + 1);
            res.push([node.val, depth]);
            if (node.right) traverse(node.right, depth + 1);
        };

        traverse(root, depth);
        return res;
    }
};
const isSymmetric = root => {
    if (!root) return true;

    const left = inorder(root.left),
          right = inorder(root.right).reverse();

    if (left.length !== right.length) return false;

    for (let i = 0; i < left.length; i++) {
        const lVal = left[i][0],
              rVal = right[i][0],
              lDepth = left[i][1],
              rDepth = right[i][1];

        if (lVal !== rVal || (lVal === rVal && lDepth !== rDepth)) {
            return false;
        }
    }
    return true;
};

File: leet_code/trees/symmetric_tree.rb
def inorder(root)
    if !root
        return []
    else
        res = []
        depth = 0
        traverse = lambda do |node, depth|
            traverse.call(node.left, depth + 1) if node.left
            res << [node.val, depth]
            traverse.call(node.right, depth + 1) if node.right
            res
        end
        traverse.call(root, depth)
    end
end


def is_symmetric(root)
    return true if !root
    left_tree, right_tree = inorder(root.left), inorder(root.right).reverse
    return false if left_tree.length != right_tree.length
    left_tree.each_with_index do |ele, idx|
      lval, rval = ele[0], right_tree[idx][0]
      ldepth, rdepth = ele[1], right_tree[idx][1]
      return false if lval != rval || (lval == rval && ldepth != rdepth)
    end
    true
end
